src/datajunction/sql/dag.py

```diff
@@ -118,18 +122,61 @@ def get_database_for_nodes(
     if not databases:
         raise Exception("No valid database was found")
 
-    active_databases = [db for db in databases if db.ping()]
-
-    if not active_databases:
-        raise Exception("No active database was found")
-
+    # if a specific database was requested, return it if it's online
     if database_id is not None:
-        for database in active_databases:
-            if database.id == database_id:
+        for database in databases:
+            if database.id == database_id and await database.do_ping():
                 return database
         raise Exception(f"Database ID {database_id} is not valid")
 
-    return sorted(active_databases, key=operator.attrgetter("cost"))[0]
+    return await get_cheapest_online_database(databases)
+
+
+async def get_cheapest_online_database(databases: Set[Database]) -> Database:
+    """
+    Return the cheapest online database.
+
+    The function will try to wait until the fastest database is pinged successfully. If
+    it's offline, it will try to wait until the second fastest, and so on. After waiting
+    for ``settings.do_ping_timeout`` it will return the fastest database that is online.
+    """
+    # sort by cheapest
+    sorted_databases = sorted(databases, key=operator.attrgetter("cost"))
+
+    # create tasks to ping all of the databases in parallel
+    pings = [asyncio.create_task(database.do_ping()) for database in sorted_databases]
+    pending = set(pings)
+
+    loop = asyncio.get_running_loop()
```

Comment:
Just one question... why not `get_event_loop()`. It looks like `get_running_loop()` will fail if there is no current loop running, and `get_event_loop()` will start one.

Reply:
Good question! In this case we're inside a coroutine (`async def`), so we know we're going to be inside an event loop, since it's the only way to call an async function as far as I know. The docs mention:

> Because this function has rather complex behavior (especially when custom event loop policies are in use), using the [get_running_loop()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_running_loop) function is preferred to [get_event_loop()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop) in coroutines and callbacks.